File: library/src/main/java/com/senierr/sehttp/cookie/SerializableCookie.java
package com.senierr.sehttp.cookie;

import android.util.Log;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.io.Serializable;

import okhttp3.Cookie;
import okhttp3.internal.Util;

/**
 * 可序列化的Cookie
 *
 * @author zhouchunjie
 * @date 2018/8/14
 */
public class SerializableCookie implements Serializable {

    private static final String TAG = SerializableCookie.class.getSimpleName();
    private static final long serialVersionUID = -8594045714036645534L;
    private static final long NON_VALID_EXPIRES_AT = -1L;

    private transient Cookie cookie;

    public SerializableCookie(Cookie cookie) {
        this.cookie = cookie;
    }

    /**
     * Cookie转字符串
     *
     * @return
     */
    public static String encode(Cookie cookie) {
        ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
        ObjectOutputStream objectOutputStream = null;

        try {
            objectOutputStream = new ObjectOutputStream(byteArrayOutputStream);
            objectOutputStream.writeObject(new SerializableCookie(cookie));
        } catch (IOException e) {
            Log.d(TAG, "IOException in encodeCookie", e);
            return null;
        } finally {
            Util.closeQuietly(objectOutputStream);
        }
        return byteArrayToHexString(byteArrayOutputStream.toByteArray());
    }

    /**
     * 字符串转Cookie
     *
     * @param encodedCookie
     * @return
     */
    public static Cookie decode(String encodedCookie) {
        byte[] bytes = hexStringToByteArray(encodedCookie);
        ByteArrayInputStream byteArrayInputStream = new ByteArrayInputStream(
                bytes);

        Cookie cookie = null;
        ObjectInputStream objectInputStream = null;
        try {
            objectInputStream = new ObjectInputStream(byteArrayInputStream);
            cookie = ((SerializableCookie) objectInputStream.readObject()).getCookie();
        } catch (IOException e) {
            Log.d(TAG, "IOException in decodeCookie", e);
        } catch (ClassNotFoundException e) {
            Log.d(TAG, "ClassNotFoundException in decodeCookie", e);
        } finally {
            Util.closeQuietly(objectInputStream);
        }
        return cookie;
    }

    /**
     * 二进制数组转十六进制字符串
     *
     * @param bytes
     * @return
     */
    private static String byteArrayToHexString(byte[] bytes) {
        StringBuilder sb = new StringBuilder(bytes.length * 2);
        for (byte element : bytes) {
            int v = element & 0xff;
            if (v < 16) {
                sb.append('0');
            }
            sb.append(Integer.toHexString(v));
        }
        return sb.toString();
    }

    /**
     * 十六进制字符串转二进制数组
     *
     * @param hexString
     * @return
     */
    private static byte[] hexStringToByteArray(String hexString) {
        int len = hexString.length();
        byte[] data = new byte[len / 2];
        for (int i = 0; i < len; i += 2) {
            data[i / 2] = (byte) ((Character.digit(hexString.charAt(i), 16) << 4) + Character
                    .digit(hexString.charAt(i + 1), 16));
        }
        return data;
    }

    private void writeObject(ObjectOutputStream out) throws IOException {
        out.writeObject(cookie.name());
        out.writeObject(cookie.value());
        out.writeLong(cookie.persistent() ? cookie.expiresAt() : NON_VALID_EXPIRES_AT);
        out.writeObject(cookie.domain());
        out.writeObject(cookie.path());
        out.writeBoolean(cookie.secure());
        out.writeBoolean(cookie.httpOnly());
        out.writeBoolean(cookie.hostOnly());
    }

    private void readObject(ObjectInputStream in) throws IOException, ClassNotFoundException {
        Cookie.Builder builder = new Cookie.Builder();
        builder.name((String) in.readObject());
        builder.value((String) in.readObject());
        long expiresAt = in.readLong();
        if (expiresAt != NON_VALID_EXPIRES_AT) {
            builder.expiresAt(expiresAt);
        }

        final String domain = (String) in.readObject();
        builder.domain(domain);
        builder.path((String) in.readObject());
        if (in.readBoolean())
            builder.secure();
        if (in.readBoolean())
            builder.httpOnly();
        if (in.readBoolean())
            builder.hostOnlyDomain(domain);
        cookie = builder.build();
    }

    public Cookie getCookie() {
        return cookie;
    }

    public void setCookie(Cookie cookie) {
        this.cookie = cookie;
    }

    @Override
    public boolean equals(Object obj) {
        return obj instanceof SerializableCookie
                && cookie.equals(((SerializableCookie) obj).getCookie());
    }

    @Override
    public int hashCode() {
        return cookie.hashCode();
    }
}

File: rxhttp/src/main/java/com/senierr/http/cookie/SerializableCookie.kt
package com.senierr.http.cookie

import android.util.Log
import com.senierr.http.util.Utils

import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.ObjectInputStream
import java.io.ObjectOutputStream
import java.io.Serializable

import okhttp3.Cookie

/**
 * 可序列化的Cookie
 *
 * @author zhouchunjie
 * @date 2018/8/14
 */
class SerializableCookie(@Transient private var cookie: Cookie? = null) : Serializable {

    private fun writeObject(out: ObjectOutputStream) {
        cookie?.let {
            out.writeObject(it.name())
            out.writeObject(it.value())
            out.writeLong(if (it.persistent()) it.expiresAt() else NON_VALID_EXPIRES_AT)
            out.writeObject(it.domain())
            out.writeObject(it.path())
            out.writeBoolean(it.secure())
            out.writeBoolean(it.httpOnly())
            out.writeBoolean(it.hostOnly())
        }
    }

    private fun readObject(inputStream: ObjectInputStream) {
        val builder = Cookie.Builder()
        builder.name(inputStream.readObject() as String)
        builder.value(inputStream.readObject() as String)
        val expiresAt = inputStream.readLong()
        if (expiresAt != NON_VALID_EXPIRES_AT) {
            builder.expiresAt(expiresAt)
        }

        val domain = inputStream.readObject() as String
        builder.domain(domain)
        builder.path(inputStream.readObject() as String)
        if (inputStream.readBoolean()) builder.secure()
        if (inputStream.readBoolean()) builder.httpOnly()
        if (inputStream.readBoolean()) builder.hostOnlyDomain(domain)
        cookie = builder.build()
    }

    override fun equals(other: Any?): Boolean {
        return other is SerializableCookie && cookie == other.cookie
    }

    override fun hashCode(): Int {
        return cookie.hashCode()
    }

    companion object {

        private val TAG = SerializableCookie::class.java.simpleName
        private const val serialVersionUID = -8594045714036645534L
        private const val NON_VALID_EXPIRES_AT = -1L

        /**
         * Cookie转字符串
         */
        fun encode(cookie: Cookie): String? {
            val byteArrayOutputStream = ByteArrayOutputStream()
            var objectOutputStream: ObjectOutputStream? = null

            try {
                objectOutputStream = ObjectOutputStream(byteArrayOutputStream)
                objectOutputStream.writeObject(SerializableCookie(cookie))
            } catch (e: IOException) {
                Log.d(TAG, "IOException in encodeCookie", e)
                return null
            } finally {
                Utils.closeQuietly(objectOutputStream)
            }
            return byteArrayToHexString(byteArrayOutputStream.toByteArray())
        }

        /**
         * 字符串转Cookie
         */
        fun decode(encodedCookie: String): Cookie? {
            val bytes = hexStringToByteArray(encodedCookie)
            val byteArrayInputStream = ByteArrayInputStream(bytes)

            var cookie: Cookie? = null
            var objectInputStream: ObjectInputStream? = null
            try {
                objectInputStream = ObjectInputStream(byteArrayInputStream)
                cookie = (objectInputStream.readObject() as SerializableCookie).cookie
            } catch (e: IOException) {
                Log.d(TAG, "IOException in decodeCookie", e)
            } catch (e: ClassNotFoundException) {
                Log.d(TAG, "ClassNotFoundException in decodeCookie", e)
            } finally {
                Utils.closeQuietly(objectInputStream)
            }
            return cookie
        }

        /**
         * 二进制数组转十六进制字符串
         *
         * @param bytes
         * @return
         */
        private fun byteArrayToHexString(bytes: ByteArray): String {
            val sb = StringBuilder(bytes.size * 2)
            for (element in bytes) {
                val v = element.toInt() and 0xff
                if (v < 16) {
                    sb.append('0')
                }
                sb.append(Integer.toHexString(v))
            }
            return sb.toString()
        }

        /**
         * 十六进制字符串转二进制数组
         *
         * @param hexString
         * @return
         */
        private fun hexStringToByteArray(hexString: String): ByteArray {
            val len = hexString.length
            val data = ByteArray(len / 2)
            for (i in 0 until len step 2) {
                data[i / 2] = ((Character.digit(hexString[i], 16) shl 4) + Character.digit(hexString[i + 1], 16)).toByte()
            }
            return data
        }
    }
}